### PR TITLE
Addition of  partition_ddf to Ktk

### PR DESCRIPTION
**Pull request by usha-nemani-by**
This PR is continuation to PR#357 , to add more changes based on the discussion.

Addition of  partition_ddf func to Ktk

- [ ] Addition of  partition_ddf  func to Ktk
- [ ] New test cases addition. 
- [ ] Changelog entry

### CODE REVIEW

**Comment by usha-nemani-by** [comment]
@fjetter Comment :

I'm still not entirely sold on this API. The repartition ratio itself was more or less a workaround I introduced back in the day to justify the repartitioning in update_dataset_from_ddf. Taking a step back, we'd be introducing two functions here which would always be called together

```
ratio_graph = calculate_repartition_ratio(..., target_nrows_per_partition, frac)
ratio = ratio_graph.compute()
new_ddf = repartition_ddf(ddf, repartition_ratio=ratio)

```
I'm wondering if we should not instead introduce one function

```
def repartition_ddf(ddf, partition_size: int):
    # We would block here
    npartitions = _calc_npartitions(...).compute()
    return ddf.repartition(npartitions=npartitions)
```

The implementation would be more or less the same but it feels to me as if the API was more user friendly. The only reason why two separate calls would be justified if we'd wanted to stick to the lazy execution.

**Comment by lr4d** [comment]
I had already approved the (previous) PR. @fjetter Can you share your thoughts?

**Comment by fjetter** [comment]
My thoughts are effectively written above. the repartition ratio is not a good api I believe, since it's rather obscure without delivering actual value to the user in terms of functionality.
I introduced the parameter back in the day on application level and figured that repartitioning is essential for shuffling but I didn't have a proper tool to describe it better and just moved the application level config to the library. However, the repartition ratio has little value compared to the plain `npartitions` of task. The value we want to offer in this PR is that we may calculate rapidly how many partitions are requested by the user. In effect we're offering a faster alternative to the existing api of dask but dask doesn't have a repartition ratio. 
Therefore, I argue that that we should rather use quantities already familiar. 
After all, the task/story we are working on at the moment is about this exact problem. Repartiton ratio is a fragile abstraction. Number of rows (or size in MB) is much more tangible and robust

**Comment by lr4d** [comment]
>  Therefore, I argue that that we should rather use quantities already familiar.

I agree

File: kartothek/io/dask/dataframe.py [hunk 1]
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
...
+    predicates: Optional[PredicatesType] = None,
+    *,
+    target_nrows_per_partition: int,
+    dataset_metadata_frac: float,
+) -> dd.core.Scalar:
+    """
+    Calculate a `repartition_ratio` according to the dataset input and `target_nrows_per_partition` (the desired
+    number of rows of each dataframe partition).
+
+
```
Comment (marco-neumann-by) [inline kartothek/io/dask/dataframe.py:620]: 1. please add a note that this function performs IO.
2. "calculate" could be read by many users as "compute on client". I would prefer something like `compute`.

File: kartothek/io/dask/dataframe.py [hunk 2]
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
...
+) -> dd.DataFrame:
+    """
+    Repartition a dask DataFrame according, dividing the current number of partitions by `repartition_ratio`.
+
+    Parameters
+    ----------
+    ddf
+        dask.Dataframe
+    repartition_ratio
+        calculated repartition ratio
```
Comment (marco-neumann-by) [inline kartothek/io/dask/dataframe.py:660]: Please linke `calculate_repartition_ratio`

File: kartothek/io/dask/dataframe.py [hunk 3]
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
...
+    mean_nrows_per_partition = expected_nrows_per_partition.mean()
+    repartition_ratio = target_nrows_per_partition / mean_nrows_per_partition
+    return repartition_ratio
+
+
+def repartition_ddf(
+    ddf: dd.DataFrame, repartition_ratio: SupportsFloat
+) -> dd.DataFrame:
+    """
+    Repartition a dask DataFrame according, dividing the current number of partitions by `repartition_ratio`.
```
Comment (marco-neumann-by) [inline kartothek/io/dask/dataframe.py:653]: What exactly is a "repartition ratio"? Neither this nor the other function explain this term.

File: tests/io/dask/dataframe/test_repartition.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    # this is somewhat arbitrary as original partition size may be non-constant
+    margin_of_error = 0.5  # proportion
+    partition_lengths = ddf.map_partitions(len).compute()
+    assert (
+        ((1 - margin_of_error) * target_nrows_per_partition <= partition_lengths)
+        & (partition_lengths <= (1 + margin_of_error) * target_nrows_per_partition)
+    ).all()
+
+
+@pytest.mark.parametrize("repartition_ratio", (3.88, 0.25, 7, 1.0))
```
Comment (marco-neumann-by) [inline tests/io/dask/dataframe/test_repartition.py:92]: where do these numbers come from?

**Review by marco-neumann-by — CHANGES_REQUESTED** [review]

File: kartothek/io/dask/dataframe.py (hunk 3 above)
Comment (marco-neumann-by) [reply kartothek/io/dask/dataframe.py:653]: I should have read the PR comments before, sorry. I see you already more or less agreed that this term is somewhat fragile / non-user-friendly.

**Comment by usha-nemani-by** [comment]
>Therefore, I argue that that we should rather use quantities already familiar.
> After all, the task/story we are working on at the moment is about this exact problem. Repartiton ratio is a fragile 
abstraction. Number of rows (or size in MB) is much more tangible and robust


Based on above discussion, I modified the functionality.
I have removed the `calculate_repartition_ratio()` function.
Modified the `repartition_ddf()` functionality  as `compute_repartition_ddf()`to compute  the repartition_ratio and compute `target_npartitions` and finally repartition the ddf.
I felt that `compute_repartition_ddf()` itself is a small function with 4 steps:
1. Collect dataset_metadata
2. compute repartition_ratio using `target_nrows_per_partition`
3. calculate `target_npartitions` using above ratio.
4. repartition the ddf and optimize it.


If I divide the function into 2 smaller funcs again, I need to send all the parameters from `compute_repartition_ddf()` to the `_calc_target_npartitions()` as well.

```
def compute_repartition_ddf(ddf, dataset_factory, target_nrows_per_partition, frac, predicates , ):
      code block here
     _calc_target_npartitions(ddf, dataset_factory, target_nrows_per_partition, frac, predicates)
  
```

I have updated the description for the same.
Please confirm if the single function looks fine.

File: tests/io/dask/dataframe/test_repartition.py (hunk 4 above)
Comment (usha-nemani-by) [reply tests/io/dask/dataframe/test_repartition.py]: They are random numbers - >1,  <1 , ==1

File: kartothek/io/dask/dataframe.py (hunk 2 above)
Comment (usha-nemani-by) [reply kartothek/io/dask/dataframe.py]: We are removing this functionality. Please review the latest code changes.

File: kartothek/io/dask/dataframe.py (hunk 1 above)
Comment (usha-nemani-by) [reply kartothek/io/dask/dataframe.py:625]: `calculate_repartion_ratio()` function is removed.

File: tests/io/dask/dataframe/test_repartition.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    "df,target_nrows_per_partition",
+    (
+        (
+            pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}),
+            97,
+        ),  # rep_ratio = 3.88
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 1),  # rep_ratio = 1.0
+    ),
+)
+def test_calculate_repartition_ratio_approximate_match(
```
Comment (lr4d) [inline tests/io/dask/dataframe/test_repartition.py:57]:
```suggestion
def test_repartition_ddf_approximate_match(
```

File: kartothek/io/dask/dataframe.py [hunk 6]
```diff
@@ -606,3 +607,53 @@ def hash_dataset(
...
+    dataset_factory
+        dataset factory
+    predicates
+        Optional , but predicates can be applied while collecting_dataset_metadata
+    target_nrows_per_partition
+        target num of rows per partition
+    dataset_metadata_frac
+        This fraction is helps in collecting metadata on a fraction(sample) of the ddf, instead of the whole ddf.
+
+
```
Comment (lr4d) [inline kartothek/io/dask/dataframe.py:639]: I was thinking if we should add `chunksize`as dask allows i.e. the uncompressed size per partition. That is probably more valuable than `target_nrows_per_partition`

File: kartothek/io/dask/dataframe.py [hunk 7]
```diff
@@ -606,3 +607,53 @@ def hash_dataset(
...
+    The input ddf is repartitioned using target_npartitions value.
+
+    Parameters
+    ----------
+    ddf
+        dask.DataFrame
+    dataset_factory
+        dataset factory
+    predicates
+        Optional , but predicates can be applied while collecting_dataset_metadata
```
Comment (lr4d) [inline kartothek/io/dask/dataframe.py:633]:
```suggestion
    predicates
        Optional , but predicates can be applied while collecting_dataset_metadata.
        Note that predicates will only be applied during query planning, predicate pushdown is not (yet) supported.
        Thus, this is only useful if providing predicates involving indexed columns.
```

File: tests/io/dask/dataframe/test_repartition.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+#     )
+#     ddf = read_dataset_as_ddf(store=store_factory, dataset_uuid="test")
+#     optimized_ddf = dask.optimize(ddf, optimizations=[dask.dataframe.optimize])[0]
+#     repartitioned_ddf = repartition_ddf(ddf, repartition_ratio=repartition_ratio)
+#     if repartition_ratio >= 1:
+#         assert len(repartitioned_ddf.dask) <= len(
+#             optimized_ddf.dask
+#         )  # assert the number of tasks in dask graph
+#     else:
+#         assert len(repartitioned_ddf.dask) > len(optimized_ddf.dask)
```
Comment (lr4d) [inline tests/io/dask/dataframe/test_repartition.py:98]: We'll want to replace this with something analogous for repartition_ddf (target_nrows or chunksize)

**Review by lr4d — CHANGES_REQUESTED** [review]

**Comment by lr4d** [comment]
> If I divide the function into 2 smaller funcs again, I need to send all the parameters from repartition_ddf() to the _calc_target_npartitions() as well.

I think it makes sense to isolate the "calculate" logic from the repartitioning logic though.

File: tests/io/dask/dataframe/test_repartition.py (hunk 8 above)
Comment (usha-nemani-by) [reply tests/io/dask/dataframe/test_repartition.py:98]: The first two testcase parameters are ddf and  target_nrows . So I was thinking, is this testcase required at all?

File: kartothek/io/dask/dataframe.py (hunk 6 above)
Comment (usha-nemani-by) [reply kartothek/io/dask/dataframe.py:641]: We should fix about this Parameter as this would be the customer facing one. 
@fjetter , any thoughts here?

**Comment by usha-nemani-by** [comment]
> > If I divide the function into 2 smaller funcs again, I need to send all the parameters from repartition_ddf() to the _calc_target_npartitions() as well.
> 
> I think it makes sense to isolate the "calculate" logic from the repartitioning logic though.

I have isolated the calculation logic. Please verify and confirm.

File: kartothek/io/dask/dataframe.py (hunk 3 above)
Comment (usha-nemani-by) [reply kartothek/io/dask/dataframe.py]: The logic is updated completely. Please review new logic and give suggestions.

**Comment by usha-nemani-by** [comment]
@fjetter / @lr4d , 
In the previous PR  (#357 ), there was a suggestion to call   `repartition_ddf()` in   [https://github.com/JDASoftwareGroup/kartothek/blob/9e4909102c0b5d95b36d25136a6479fb71c644ab/kartothek/io/dask/dataframe.py#L336](url)  

If that is the case, do I  need to  remove the `repartition_ratio` from input params, and change the signature of that function?

[https://github.com/JDASoftwareGroup/kartothek/blob/9e4909102c0b5d95b36d25136a6479fb71c644ab/kartothek/io/dask/dataframe.py#L431](url)

**Comment by lr4d** [comment]
After discussion involving Usha, Florian and I, it was decided to close this until we agree on an API interface for this functionality

**Pull request closed without merging**